Some (mostly backported) refactor

Kaffeine opened the pull request:
Port some cleanup and refactoring from the upstream and add some custom refactoring.

Changes overview:
- Backported angle() and direction() functions (upstream commit https://github.com/ddnet/ddnet/commit/7735557aa5c6fb81daa724cce3caa109d1c76c11)
- Backported vector math cleanup (remove some pointless templates, refactor other functions; upstream commit https://github.com/ddnet/ddnet/commit/bdaabc10bcb04e28cf650bb489b87fe8a0e03dee)
- `ddracechat.h` (used by both client and server) moved to a shared dir next to `ddracecommands.h`
- Backport `IGameController::DoActivityCheck()`
- Backport `IGameController::DoTeamChange()`
- Backport `IGameController::OnPlayerDisconnect()`
- Refactor Emote-related code (decouple the code and hide the details in CPlayer class)
- Fix: restore default emote on the emote override ended over.
- CCharacter::HandleTiles(): Move a part of DDRace-specific code to the DDRace GameController
- Add `IGameController::GetMaskForPlayerWorldEvent()` to port classes away from direct `CGameControllerDDRace/CGameTeams` usage.

Result overview:
- CPlayer does not depend on CGameControllerDDRace nor CGameTeams anymore
- CCharacter does not depend on CGameControllerDDRace anymore

I'm not sure that `GetMaskForPlayerWorldEvent()` is a good name for that. The idea of the method is to abstract the users of `CreateExplosion`, `CreatePlayerSpawn`, `CreateDeath`, `CreateSound`, and other world events from thinking about the player team.

I checked that the client and server work (didn't do deep testing though).

## Checklist

- [x] Tested the change ingame
- Provided screenshots if it is a visual change (no visual changes)
- Tested in combination with possibly related configuration options (no related options)
- [ ] Written a unit test if it works standalone, system.c especially
- [x] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

def- commented:
On a quick look this seems fine to me. Other opinions?

heinrich5991 commented on src/base/vmath.h:
```diff
@@ -87,17 +87,23 @@ inline T length(const vector2_base<T> &a)
+{
+	return atan2f(a.y, a.x);
+}
+
```
This change has the potential to affect gameplay, but I'm fine with trying it out and seeing if anyone notices anything.

Kaffeine replied:
Yeah, I was noticed that too. Some docs say that the second arg should not be zero and can trigger "domain error". But the upstream did it so here we are.

heinrich5991 replied:
https://en.cppreference.com/w/c/numeric/math/atan2

It says that there's only a problem if x and y are both zero. cppreference.com is fairly authorative on C/C++ standard.

heinrich5991 replied:
I was more talking about the fact that you replaced our manual angle calculation with the more correct atan2f. This can change how stuff is rounded, however, so it has the potential to affect real physics.

Kaffeine replied:
>This change has the potential to affect gameplay

I restored the previous GetAngle() implementation. The main goal here is to have the same API (e.g. function name and the header file).

heinrich5991 commented on src/base/vmath.h:
```diff
@@ -162,8 +159,8 @@ class vector3_base
-	vector3_base() = default;
-	vector3_base(T nx, T ny, T nz)
+	vector3_base() {}
+	vector3_base(float nx, float ny, float nz)
```
This looks wrong.

Kaffeine replied:
You're absolutely right! During the conflict solving I had to redo some T->float replacement for `length`, `angle`, and other methods.

Thank you for saving me from a regression. You have a sharp look. :+1: 

I fixed this locally but I'll wait for further comments to address them all at once.

heinrich5991 commented on src/base/vmath.h:
```diff
@@ -222,22 +219,10 @@ class vector3_base
-	float c = cosf(angle);
-	return vector2_base<T>((T)(c * a.x - s * a.y), (T)(s * a.x + c * a.y));
-}
+	operator const T *() { return &x; }
```
I think we deliberately removed this conversion. We saw nothing good coming from it, but it has the possibility of programming errors.

Kaffeine replied:
You're right we removed this from ddnet/master: f83d528f5d7b1ef9fe2709fa04ca44c16f589cc9. I'll remove it too.

heinrich5991 commented on src/game/client/gameclient.cpp:
```diff
@@ -2177,17 +2177,14 @@ int CGameClient::IntersectCharacter(vec2 HookPos, vec2 NewPos, vec2 &NewPos2, in
 
-		vec2 ClosestPoint;
-		if(closest_point_on_line(HookPos, NewPos, Position, ClosestPoint))
+		vec2 ClosestPoint = closest_point_on_line(HookPos, NewPos, Position);
```
What happens if `HookPos == NewPos`? The old code caught that. Same with the other instances of `closest_point_on_line`. I don't think this change to `closest_point_on_line` is sensible.

Kaffeine replied:
I restored the previous `closest_point_on_line()` implementation.

heinrich5991 commented on src/game/server/gamecontroller.h:
```diff
@@ -78,6 +78,45 @@ class IGameController
 
+	// event
+	/*
+		Function: on_CCharacter_death
```
Function name is wrong. Same below.

Kaffeine replied:
The commit backported as-is. I'll fix it nonetheless.

heinrich5991 commented on src/game/server/gamecontroller.h:
```diff
@@ -78,6 +78,45 @@ class IGameController
+		Returns:
+			bool?
+	*/
+	//virtual bool OnEntity(int Index, vec2 Pos);
```
Code that is commented out can be removed.

heinrich5991 commented on src/game/server/gamecontroller.h:
```diff
@@ -114,6 +114,9 @@ class IGameController
 	virtual bool OnEntity(int Index, vec2 Pos, int Layer, int Flags, int Number = 0);
 
+	void OnPlayerDisconnect(class CPlayer *pPlayer, const char *pReason);
+	// void OnPlayerInfoChange(class CPlayer *pPlayer);
```
Code that is commented out should be removed.

Kaffeine replied:
I left it here because we have commented code that calls this: https://github.com/ddnet/ddnet/blob/master/src/game/server/player.cpp#L656

I'll remove this here and I cleanup CPlayer code if you want.

heinrich5991 commented on src/game/server/gamecontroller.cpp:
```diff
@@ -468,6 +468,11 @@ void IGameController::OnCharacterSpawn(class CCharacter *pChr)
 	pChr->GiveWeapon(WEAPON_GUN);
 }
 
+void IGameController::HandleCharacterTiles(CCharacter *, int)
```
This is unusual for DDNet code, please name the parameters even if you don't use them.

Kaffeine replied:
We'll get "unused variables" warnings in that case. In other frameworks there are some UNUSED() macros; I can introduce such one and use it here. Or I can leave warnings.

heinrich5991 replied:
Do we get unused variables warnings? I thought we didn't have such a warning enabled.

def- replied:
If you get the warning then write the name in a comment with`/* name */`

heinrich5991 commented on src/game/server/gamecontroller.cpp:
```diff
@@ -657,6 +657,12 @@ int IGameController::ClampTeam(int Team)
 	return 0;
 }
 
+int64 IGameController::GetMaskForPlayerWorldEvent(int, int ExceptID)
```
Please name all parameters.

Kaffeine replied:
Done

heinrich5991 reviewed (changes requested):
Rest looks good. I went commit-by-commit and marked some places where I was unsatisfied with the solution.

1) Make sure not to modify game physics code.
2) Code that is commented out should be deleted.
3) We name all parameters in DDNet code.

Kaffeine commented:
v2:
- Removed vector2_base operator const T *() (intentionally removed by commit f83d528f5d7b1ef9fe2709fa04ca44c16f589cc9)
- Fixed accidentical vector3_base constructor arguments (were `float`).
- Name all method arguments (in IGameController::HandleCharacterTiles(), IGameController::GetMaskForPlayerWorldEvent())
- Removed some added commented-out code
- Fixed method names in the backported comments in gamecontroller.h
- vmath.h: `angle()` (ex. `GetAngle()`) kept as-is (not as the upstream version)
- vmath.h: closest_point_on_line() kept as-is (not as the upstream version)

Kaffeine commented:
Oh. Check-style failed on `datasrc/network.py`. I have no idea what could be wrong. My diff to the yesterday "v1" version https://github.com/ddnet/ddnet/commit/fbb690ec7fd164b992a556ee5abd23e0fbf00316:

```
$ git diff fbb690e..9bc562e --stat
 src/base/vmath.h                         | 59 +++++++++++++++++++++++++++++++++++++----------------------
 src/game/client/components/items.cpp     |  5 +++--
 src/game/client/gameclient.cpp           | 15 +++++++++------
 src/game/client/prediction/gameworld.cpp | 30 ++++++++++++++++++------------
 src/game/gamecore.cpp                    | 17 ++++++++++-------
 src/game/server/gamecontroller.cpp       |  4 ++--
 src/game/server/gamecontroller.h         | 21 +++------------------
 src/game/server/gameworld.cpp            | 32 +++++++++++++++++++-------------
 8 files changed, 101 insertions(+), 82 deletions(-)
```

(mostly `closest_point_on_line` and the comments)

heinrich5991 commented:
See #3507, not your fault.

Kaffeine commented:
The PR is ready for (re)review and maybe merge.

Learath2 commented:
I like all of these. I have one question: What is the motivation behind `GameController::HandleCharacterTiles`? 

Structurally the more sane approach to this would be the collision code calling into the entities when they touch a tile, which would be a rather large rework. Until that is there really much value of moving this around?

Kaffeine commented:
> What is the motivation behind `GameController::HandleCharacterTiles`?

The main motivation is to get rid of `&((CGameControllerDDRace *)GameServer()->m_pController)->` references from the codebase and make it possible to base another mod on top of DDNet without commenting-out a lot of racing-related code here, there, and somewhere else.

I have a (legacy) mod codebase on the upstream 0.6 version and I tried to reengineer it on top of upstream 0.7. Now I'm backporting the 0.7 stuff to DDNet to make the port easier, but the task is still harder (than the base on the upstream) because of the racing code spread through the codebase in both expected and very unexpected places.

I'm OK with (/would not care about) the extra tiles such as accelerators, doors, and all other stuff that extends the base game physics.

>Structurally the more sane approach to this would be the collision code calling into the entities when they touch a tile, which would be a rather large rework.

I also tried to add two methods:
```
virtual void OnCharacterTouchEntity(class CCharacter *pChr, int TileIndex);
virtual void OnCharacterCloseToEntity(class CCharacter *pChr, int TileIndex);
```
This results in much better code, like:
```diff
-if(((m_TileIndex == TILE_FINISH) 
-    || (m_TileFIndex == TILE_FINISH) 
-    || FTile1 == TILE_FINISH 
-    || FTile2 == TILE_FINISH 
-    || FTile3 == TILE_FINISH 
-    || FTile4 == TILE_FINISH 
-    || Tile1 == TILE_FINISH 
-    || Tile2 == TILE_FINISH 
-    || Tile3 == TILE_FINISH 
-    || Tile4 == TILE_FINISH) && PlayerDDRaceState == DDRACE_STARTED)
+if((TileIndex == TILE_FINISH) && PlayerDDRaceState == DDRACE_STARTED)
```
but then I realized that such code can introduce regressions/bugs on maps that rely on the order of the checks.
I can restore the code but the reorder can be dangerous.

Kaffeine commented:
> Structurally the more sane approach to this would be the collision code calling into the entities when they touch a tile, which would be a rather large rework

@Learath2: This is how it can look like:
https://github.com/InfectionDust/ddnet/compare/ddnet-refactor-base...InfectionDust:ddnet-refactor-handle-tiles

No much rework and I like it very much; it was my first approach. But I should repeat: it can trigger bugs because of reordering. E.g.:

```
int FTileIndex = TILE_START;
int TileIndex = TILE_FINISH;
```
The old code would process `TILE_START` and then `TILE_FINISH`. The new code will process `TileIndex` (and thus `TILE_FINISH`) before `FTileIndex` (`TILE_START`).

Learath2 commented:
I see what you mean. We should merge as-is for now, a more comprehensive rework or reordering requires more testing and thinking :)

Kaffeine commented:
> I see what you mean. We should merge as-is for now, a more comprehensive rework or reordering requires more testing and thinking :)

I think that a map should not rely on the order of the checks. It makes little sense to place controversial or order-dependent tiles at the same index. The order should be treated as undefined, like [C++ undefined behavior](https://en.cppreference.com/w/cpp/language/ub). But probably we still should care about those who rely on our "UB" or at least we should think twice.

> We should merge as-is for now

Probably :).

> How about the checklist, at a first glance it looks like it only moved things around but it'd be nice to know that you've tested atleast the basics, starting/finishing a race, etc.

I will do some testing tomorrow — I just wanted some ACK before spending my time on the checks (if I would get some negative acknowledgment it'll be better for me to not spend another 30-40 minutes on the server setup (I don't have such a host setup for ten years)).

Kaffeine commented:
@heinrich5991, are you OK with the code now (see [v2 details](https://github.com/ddnet/ddnet/pull/3502#issuecomment-759622381)), so it makes sense to test it?

Learath2 commented:
> I think that a map should not rely on the order of the checks. It makes little sense to place controversial or order-dependent tiles at the same index. The order should be treated as undefined, like [C++ undefined behavior](https://en.cppreference.com/w/cpp/language/ub). But probably we still should care about those who rely on our "UB" or at least we should think twice.

In an ideal world, yes but ddnet has a loong history and the mappers have mostly been unchecked for most of it. So any and all physics checks need quite a bit of scrutiny to make sure we don't break ranks or maps.

Order is veeery unlikely to break anything, but it'd still make me less certain about merging this

heinrich5991 reviewed (approved):
>@heinrich5991, are you OK with the code now (see v2 details), so it makes sense to test it?

@Kaffeine I'm okay with these changes now. :) It makes sense to test it now.

Kaffeine commented:
I tested the changes and found a terrible bug due to one missing line. I didn't call the `IGameController::OnCharacterSpawn(pChr);` in `CGameControllerDDRace::OnCharacterSpawn(CCharacter *pChr)` thus the Character had no HP and no ammo (the char had full armor nonetheless).

I fixed the commit and think that now PR is ready to land.

Learath2 reviewed (approved):
LGTM, thanks for some much needed cleanup

bors r+

C0D3D3V commented:
Is there a real reason for `GameController::GetMaskForPlayerWorldEvent()`
Or are you ok with it if I remove it again and replace it with `m_pCharacter->TeamMask()`
for all of these:
https://github.com/ddnet/ddnet/commit/c28bd693d3b8782a3b8ef824aae0e6c70ecb5bb7

Kaffeine commented:
The idea is to decouple the classes. Why would you want to couple them again?
The less classes know about each other (to some reasonable degree of course), the easier it is to change them.

Personally I don't see `teeworlds` as a good upstream for the mods, I would prefer to cleanup `ddnet` codebase and use it as the upstream instead. I would prefer to have DDNet gameplay related features concentrated in a few classes, rather then spread through the entire codebase.

What's wrong with `GameController::GetMaskForPlayerWorldEvent()`? It is up to the game controller to decide if the game mode has teams with separated events. Some game controllers need to mask the events, some do not. How it is the base `CPlayer` business?

C0D3D3V commented:
It is the same method just a longer name... The dependency to the character is from nature already there. But I do not have something against it, if you thing any mod can do something good with that.

C0D3D3V commented:
The method only felt into my eye, as I cleaned up the use of Team()->TeamMask(...)
here https://github.com/ddnet/ddnet/pull/5063/commits/e73f125093a0d691c3c2dff8030eb8b0c67aa4c3 and https://github.com/ddnet/ddnet/pull/5063/commits/e73f125093a0d691c3c2dff8030eb8b0c67aa4c3

thats all. If this helps other mods to write independent code then I think it's good.

Kaffeine commented:
I see, the `TeamMask()` is now directly available in `CCharacter`, so at least we won't have to put back `#include <game/server/teams.h>`.
Okay, I think this also works for me.
I tried to make it easier to entirely remove `CGameTeams`. With your change the modders can cut it off in the single place in `CCharacter`, so this is good enough. And also hm, now I do *not* think that `CGameTeams` is that huge and bad. Previously I though that it is necessary to remove it if it is not used (in the modified gameplay).
---
Side note: personally I think that GameController is not enough to do a significant modification. Ideally I see us having basic gameplay in CPlayer / CCharacter / CGameController (which is actually `IGameController`), and then we can encapsulate the mod gameplay into C<Mod>Player / C<Mod>Character / C<Mod>GameController using methods override.

I actually do this in a mod I currently maintain. I added `virtual CPlayer *IGameController::CreatePlayer(int ClientID)` and `CInfClassGameController::CreatePlayer() override`  returns `CInfClassPlayer` instance. In its turn,  `CInfClassPlayer` overrides `void CPlayer:;TryRespawn()` and constructs `CInfClassCharacter` instead.

The basic game logic works as is, all gametype-specific modifications are done via virtual methods overload. The mod classes know about the mod game controller, so there is no ugly cast (like `&((CGameControllerDDRace *)GameServer()->m_pController)->` partially removed in this PR).
